src/shared/s3_manager.py

```diff
@@ -238,3 +271,29 @@ def write_local_metadata(self, metadata: dict | None = None, meta_type: str | No
             metadata=metadata,
             meta_type=meta_type,
         )
+
+    # Lockfile management
+    def request_or_validate_lock(self, lock_id: str | None = None):
+        try:
+            lock = functions.get_s3_json_as_dict(bucket=self.s3_bucket_name, key=self.lockfile)
+            if lock_id != lock["id"]:
+                raise errors.AggregatorStudyProcessingError
+
+        except botocore.exceptions.ClientError:
+            # if the lockfile doesn't exist, we'll make one
+            lock_id = str(uuid.uuid4())
+            lock = {"id": lock_id, "uploaded_at": datetime.datetime.now(datetime.UTC)}
+            self.s3_client.put_object(
+                Bucket=self.s3_bucket_name,
+                Key=self.lockfile,
+                Body=json.dumps(lock, default=str, indent=2),
+            )
+            sqs_client = boto3.client("sqs")
+            sqs_client.send_message(
+                QueueUrl=os.environ.get("QUEUE_LOCKFILE_CLEANUP"),
+                MessageBody=json.dumps({"site": self.site, "study": self.study}),
+            )
```

Comment:
Might be worth sending this off before writing the lockfile, so that if you get interrupted between these two steps, you don't have a lingering lockfile.